src/solver/clause.rs

```diff
@@ -230,6 +234,67 @@ impl Clause {
         )
     }
 
+    fn conditional(
+        parent_id: VariableId,
+        requirement: Requirement,
+        condition: VersionSetId,
+        decision_tracker: &DecisionTracker,
+        requirement_candidates: impl IntoIterator<Item = VariableId>,
+        condition_candidates: impl IntoIterator<Item = VariableId>,
+    ) -> (Self, Option<[Literal; 2]>, bool) {
+        assert_ne!(decision_tracker.assigned_value(parent_id), Some(false));
+
+        let mut condition_candidates = condition_candidates.into_iter().peekable();
+        let condition_first_candidate = condition_candidates
+            .peek()
+            .copied()
+            .expect("no condition candidates");
+        let mut requirement_candidates = requirement_candidates.into_iter().peekable();
+
+        match condition_candidates
+            .find(|&condition_id| decision_tracker.assigned_value(condition_id) == Some(true))
+        {
+            Some(_) => {
+                // Condition is true, find first requirement candidate not set to false
+                if let Some(req_candidate) = requirement_candidates
+                    .find(|&req_id| decision_tracker.assigned_value(req_id) != Some(false))
+                {
+                    (
+                        Clause::Conditional(parent_id, condition, requirement),
+                        Some([parent_id.negative(), req_candidate.positive()]),
+                        false,
+                    )
+                } else {
+                    // No valid requirement candidate, use first condition candidate and mark conflict
+                    (
+                        Clause::Conditional(parent_id, condition, requirement),
+                        Some([parent_id.negative(), condition_first_candidate.negative()]),
+                        true,
+                    )
+                }
+            }
+            None => {
+                // No true condition, look for unset condition
+                if let Some(unset_condition) = condition_candidates.find(|&condition_id| {
+                    decision_tracker.assigned_value(condition_id) != Some(false)
+                }) {
+                    (
+                        Clause::Conditional(parent_id, condition, requirement),
+                        Some([parent_id.negative(), unset_condition.negative()]),
+                        false,
+                    )
+                } else {
+                    // All conditions false
+                    (
+                        Clause::Conditional(parent_id, condition, requirement),
+                        None,
+                        false,
+                    )
+                }
+            }
+        }
+    }
```

Comment:
I think this code can be simplified. We need to do the following things:

We need to find two literals (either from the parent, the condition or the requirement) which are either undecided or set to true. Since the assertion at the top already indicates that the parent meets that condition you only need to find a single literal that meets that condition.

If no such literal can be found, e.g. all the literals are false, it indicates there is a conflict and you should select a random literal (just the first for example).

There is one edge case:
- If there are no condition candidates. That is an assertion. If we would otherwise add the clause it would just be a requires clause (e.g. ~A v D1 v ... D2) which is not what we want.